AutoCollection/Statsbeat.ts

```diff
@@ -0,0 +1,263 @@
+    }
+
+    public enable(isEnabled: boolean) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._getCustomProperties();
+            this._isInitialized = true;
+        }
+        if (isEnabled) {
+            if (!this._handle) {
+                this._lastRequests = {
+                    totalRequestCount: this._totalRequestCount,
+                    time: +new Date
+                };
+                this._handle = setInterval(() => {
+                    this.trackStatsbeatMetrics().catch((error) => {
+                        // Failed to send Statsbeat
+                        Logging.info(Statsbeat.TAG, error);
+                    });
+                }, this._collectionInterval);
+                this._handle.unref(); // Allow the app to terminate even while this loop is going on
+            }
+        } else {
+            if (this._handle) {
+                clearInterval(this._handle);
+                this._handle = null;
+            }
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    public isEnabled() {
+        return this._isEnabled;
+    }
+
+    public setCodelessAttach() {
+        this._attach = Constants.StatsbeatAttach.codeless;
+    }
+
+    public addFeature(feature: Constants.StatsbeatFeature) {
+        this._features |= feature;
+    }
+
+    public removeFeature(feature: Constants.StatsbeatFeature) {
+        this._features &= ~feature;
+    }
+
+    public addInstrumentation(instrumentation: Constants.StatsbeatInstrumentation) {
+        this._instrumentations |= instrumentation;
+    }
+
+    public removeInstrumentation(instrumentation: Constants.StatsbeatInstrumentation) {
+        this._instrumentations &= ~instrumentation;
+    }
+
+    public countRequest(duration: number, success: boolean) {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._totalRequestCount++;
+        this._intervalRequestExecutionTime += duration;
+        if (success === false) {
+            this._totalFailedRequestCount++;
+        }
+        else {
+            this._totalSuccesfulRequestCount++;
+        }
+    }
+
+    public countException() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._exceptionCount++;
+    }
+
+    public countThrottle() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._throttleCount++;
+    }
+
+    public countRetry() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._retryCount++;
+    }
+
+    public async trackStatsbeatMetrics() {
+        this._getResourceProvider(async () => {
+            this._trackRequestDuration();
+            this._trackRequestsCount();
+            await this._sendStatsbeats();
+        });
+    }
+
+    private _trackRequestDuration() {
+        var lastRequests = this._lastRequests;
```

Comment:
nit: Technically, (but probably unlikely) someone could call trackStatsbeatMetic() (as it's public) before its really enabled which would mean that this._lastRequests will be ```undefined`` and then cause bad things to happen here